File: packages/qnsi/src/tenant.ts
/**
 * QNSP Tenant - tenant CRUD, crypto-policy management, current-health,
 * current-quotas. Wraps `apps/tenant-service` (`/tenant/v1`).
 */

import type { Internal, RequestOptions } from "./_internal.js";

const PATH_PREFIX = "/proxy/tenant/v1";

export interface TenantSignature {
	readonly provider: string;
	readonly algorithm: string;
	readonly value: string;
	readonly publicKey: string;
}

export interface TenantSecurityEnvelope {
	readonly controlPlaneTokenSha256: string | null;
	readonly pqcSignatures: readonly TenantSignature[];
	readonly hardwareProvider: string | null;
	readonly attestationStatus: string | null;
	readonly attestationProof: string | null;
}

export interface CreateTenantRequest {
	readonly name: string;
	readonly slug: string;
	readonly plan?: string;
	readonly region?: string;
	readonly complianceTags?: readonly string[];
	readonly hsmMode?: "none" | "supported" | "required";
	readonly metadata?: Record<string, unknown>;
	readonly domains?: readonly {
		readonly domain: string;
		readonly verified?: boolean;
	}[];
	readonly security: TenantSecurityEnvelope;
	readonly signature?: TenantSignature;
}

export interface UpdateTenantRequest {
	readonly name?: string;
	readonly plan?: string;
	readonly status?: "active" | "suspended" | "pending" | "deleted";
	readonly complianceTags?: readonly string[];
	readonly hsmMode?: "none" | "supported" | "required";
	readonly metadata?: Record<string, unknown>;
	readonly security: TenantSecurityEnvelope;
	readonly signature?: TenantSignature;
}

export class TenantClient {
	constructor(private readonly internal: Internal) {}

	createTenant(req: CreateTenantRequest, opts?: Pick<RequestOptions, "idempotencyKey">) {
		return this.internal.request("POST", `${PATH_PREFIX}/tenants`, req, opts);
	}

	getTenant(tenantId: string) {
		return this.internal.request("GET", `${PATH_PREFIX}/tenants/${tenantId}`);
	}

	updateTenant(
		tenantId: string,
		body: UpdateTenantRequest,
		opts?: Pick<RequestOptions, "idempotencyKey">,
	) {
		return this.internal.request("PATCH", `${PATH_PREFIX}/tenants/${tenantId}`, body, opts);
	}

	listTenants(query?: RequestOptions["query"]) {
		return this.internal.request("GET", `${PATH_PREFIX}/tenants`, undefined, { query });
	}

	getCryptoPolicy(tenantId: string) {
		return this.internal.request("GET", `${PATH_PREFIX}/tenants/${tenantId}/crypto-policy`);
	}

	upsertCryptoPolicy(
		tenantId: string,
		body: Record<string, unknown>,
		opts?: Pick<RequestOptions, "idempotencyKey">,
	) {
		return this.internal.request(
			"PUT",
			`${PATH_PREFIX}/tenants/${tenantId}/crypto-policy`,
			body,
			opts,
		);
	}

	getCurrentHealth(tenantId: string) {
		return this.internal.request("GET", `${PATH_PREFIX}/health/current`, undefined, {
			query: { tenantId },
		});
	}

	getCurrentQuotas(tenantId: string) {
		return this.internal.request("GET", `${PATH_PREFIX}/quotas/current`, undefined, {
			query: { tenantId },
		});
	}
}

File: sdks/rust/qnsi/src/tenant.rs
//! QNSI Tenant — tenant CRUD, crypto-policy management, sub-tenant
//! onboarding, and quota / health introspection.

use reqwest::Method;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::sync::Arc;

use crate::activation::Activation;
use crate::errors::Error;
use crate::service_client::ServiceClient;

const PATH_PREFIX: &str = "/proxy/tenant/v1";

#[derive(Clone)]
pub struct Client {
    sc: ServiceClient,
}

impl Client {
    pub fn new(activation: Arc<Activation>, http: reqwest::Client) -> Self {
        Self { sc: ServiceClient::new(activation, http, PATH_PREFIX) }
    }

    pub async fn create_tenant(&self, req: CreateTenantRequest, idempotency_key: Option<&str>) -> Result<Value, Error> {
        self.sc.request(Method::POST, "/tenants", Some(&req), None, idempotency_key).await
    }

    pub async fn get_tenant(&self, tenant_id: &str) -> Result<Value, Error> {
        self.sc.request::<()>(Method::GET, &format!("/tenants/{}", tenant_id), None, None, None).await
    }

    pub async fn update_tenant(&self, tenant_id: &str, body: serde_json::Map<String, Value>, idempotency_key: Option<&str>) -> Result<Value, Error> {
        self.sc.request(Method::PATCH, &format!("/tenants/{}", tenant_id), Some(&Value::Object(body)), None, idempotency_key).await
    }

    pub async fn list_tenants(&self, query: &[(&str, String)]) -> Result<Value, Error> {
        self.sc.request::<()>(Method::GET, "/tenants", None, Some(query), None).await
    }

    pub async fn get_crypto_policy(&self, tenant_id: &str) -> Result<Value, Error> {
        self.sc.request::<()>(Method::GET, &format!("/tenants/{}/crypto-policy", tenant_id), None, None, None).await
    }

    pub async fn upsert_crypto_policy(&self, tenant_id: &str, body: serde_json::Map<String, Value>, idempotency_key: Option<&str>) -> Result<Value, Error> {
        self.sc.request(Method::PUT, &format!("/tenants/{}/crypto-policy", tenant_id), Some(&Value::Object(body)), None, idempotency_key).await
    }

    pub async fn get_current_health(&self, tenant_id: &str) -> Result<Value, Error> {
        self.sc.request::<()>(Method::GET, &format!("/tenants/{}/health", tenant_id), None, None, None).await
    }

    pub async fn get_current_quotas(&self, tenant_id: &str) -> Result<Value, Error> {
        self.sc.request::<()>(Method::GET, &format!("/tenants/{}/quotas", tenant_id), None, None, None).await
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateTenantRequest {
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub slug: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tier: Option<String>,
    #[serde(rename = "parentTenantId", skip_serializing_if = "Option::is_none")]
    pub parent_tenant_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<serde_json::Map<String, Value>>,
}
